### PR TITLE
Softmax is migrated (wip)

### PR DESCRIPTION
**Pull request by jacekpie**

### CODE REVIEW

File: softmax/src/lib.rs [hunk 1]
```diff
@@ -1,82 +1,191 @@
...
 }
 
-impl Transform<Tensor<f32>> for Softmax {
-    type Output = Tensor<f32>;
+impl proc_block_v1::ProcBlockV1 for ProcBlockV1 {
+    fn register_metadata() {
+        let metadata = Metadata::new("Softmax", env!("CARGO_PKG_VERSION"));
+        metadata.set_description(env!("CARGO_PKG_DESCRIPTION"));
+        metadata.set_repository(env!("CARGO_PKG_REPOSITORY"));
+        metadata.set_homepage(env!("CARGO_PKG_HOMEPAGE"));
```
Comment (Michael-F-Bryan) [inline softmax/src/lib.rs:54]: Can you set the `description` and `homepage` fields under the `[package]` section in `Cargo.toml`? That way the text and links shown in Forge won't be empty.

File: softmax/src/lib.rs [hunk 2]
```diff
@@ -1,82 +1,191 @@
...
+        soft_max.add_hint(&hint);
+        metadata.add_output(&soft_max);
+
+        register_node(&metadata);
+    }
+
+    fn graph(id: String) -> Result<(), GraphError> {
+        let ctx = GraphContext::for_node(&id).ok_or_else(|| {
+            GraphError::Other("Unable to get the graph context".to_string())
+        })?;
```
Comment (Michael-F-Bryan) [inline softmax/src/lib.rs:85]: We found we were handling this sort of error so frequently we added a specific `GraphError::MissingContext` variant for it.

```suggestion
        let ctx = GraphContext::for_node(&id).ok_or(GraphError::MissingContext)?;
```

File: softmax/src/lib.rs [hunk 3]
```diff
@@ -1,82 +1,191 @@
...
-                supported_shapes(&[ElementType::Float32], Dimensions::Dynamic);
-            output.add_hint(&hint);
-            metadata.add_output(&output);
-
-            register_node(&metadata);
-        }
+    fn kernel(id: String) -> Result<(), KernelError> {
+        let ctx = KernelContext::for_node(&id).ok_or_else(|| {
+            KernelError::Other("Unable to get the kernel context".to_string())
+        })?;
```
Comment (Michael-F-Bryan) [inline softmax/src/lib.rs:101]: Same as `GraphError::MissingContext`.

```suggestion
        let ctx = KernelContext::for_node(&id).ok_or(KernelError::MissingContext)?;
```

File: softmax/src/lib.rs [hunk 4]
```diff
@@ -1,82 +1,191 @@
...
+        Ok(())
     }
 }
 
-#[cfg(test)]
 mod tests {
+
     use super::*;
-    use alloc::vec;
+
```
Comment (Michael-F-Bryan) [inline softmax/src/lib.rs:141]: Can we add a test for empty tensors?

**Review by Michael-F-Bryan — COMMENTED** [review]
Overall I'm pretty happy with this proc-block. There are a couple suggestions, but nothing is critical.

Once the comments have been addressed, I reckon this will be good to merge :+1:
